torch/distributed/elastic/timer/file_based_local_timer.py

```diff
@@ -68,69 +64,67 @@ class FileTimerClient(TimerClient):
+        timeout_sec: int,
+        scope_id: str,
+        signal=(signal.SIGKILL if sys.platform != "win32" else
+                signal.CTRL_C_EVENT)) -> None:  # type: ignore[attr-defined]
         super().__init__()
         self._file_path = file_path
+        self._timeout_sec = timeout_sec
         self.signal = signal
-
-    def _open_non_blocking(self) -> Optional[io.TextIOWrapper]:
-        try:
-            fd = os.open(self._file_path, os.O_WRONLY | os.O_NONBLOCK)
-            return os.fdopen(fd, "wt")
-        except Exception:
-            return None
-
-    def _send_request(self, request: FileTimerRequest) -> None:
-        # The server may have crashed or may haven't started yet.
-        # In such case, calling open() in blocking model blocks the client.
-        # To avoid such issue, open it in non-blocking mode, and an OSError will
-        # be raised if the server is not there.
-        file = self._open_non_blocking()
-        if file is None:
-            raise BrokenPipeError("Could not send the FileTimerRequest because FileTimerServer is not available.")
-        with file:
+        log.info(
+            "Starting %s..."
+            " file_path=%s,"
+            " timeout=%s,"
+            " signal=%s",
+            type(self).__name__,
+            self._file_path,
+            self._timeout_sec,
+            self.signal,
+        )
+        with open(self._file_path, "w") as file:
+            # FileTimerRequest at init is used to write client level
+            # metadata once to watchdog file, later acquire will not
+            # write duplicate data
+            request = FileTimerRequest(
+                worker_pid=os.getpid(),
+                scope_id=scope_id,
+                expiration_time=timeout_sec,
+                signal=signal,
+            )
             json_request = request.to_json()
-            # Write request with no greater than select.PIPE_BUF is guarantee to be atomic.
-            if len(json_request) > select.PIPE_BUF:
-                raise RuntimeError(
-                    f"FileTimerRequest larger than {select.PIPE_BUF} bytes "
-                    f"is not supported: {json_request}"
-                )
             file.write(json_request + "\n")
 
     def acquire(self, scope_id: str, expiration_time: float) -> None:
-        self._send_request(
-            request=FileTimerRequest(
-                worker_pid=os.getpid(),
-                scope_id=scope_id,
-                expiration_time=expiration_time,
-                signal=self.signal
-            ),
-        )
+        if not os.path.isfile(self._file_path):
+            raise FileNotFoundError("Could not send the FileTimerRequest because FileTimerServer is not available.")
+        # updating watchdog file modified time to current time as heartbeat
+        # avoid writing duplicate FileTimerRequest, expiration time is
+        # calcuated using 'modified time + timeout'
+        os.utime(self._file_path)
```

Comment:
I don't love using utime like this, wondering if it's better to atomically write the primary file since it's a bit more obvious what's going on + more flexibility.

Reply:
as we are not reading the file contents, only checking the last modified time for heartbeat. how does atomic writing will help? what scenarios it may not work?